pkg/controllers/dnspolicy/dns_helper.go

```diff
@@ -355,3 +437,29 @@ func (r *dnsHelper) deleteDNSRecordForListener(ctx context.Context, owner metav1
 func isWildCardListener(l gatewayv1beta1.Listener) bool {
 	return strings.HasPrefix(string(*l.Hostname), "*")
 }
+
+func (dh *dnsHelper) getDNSHealthCheckProbes(ctx context.Context, gateway *gatewayv1beta1.Gateway, dnsPolicy *v1alpha1.DNSPolicy) ([]*v1alpha1.DNSHealthCheckProbe, error) {
+	list := &v1alpha1.DNSHealthCheckProbeList{}
+	if err := dh.List(ctx, list, &client.ListOptions{
+		LabelSelector: labels.SelectorFromSet(commonDNSRecordLabels(client.ObjectKeyFromObject(gateway), client.ObjectKeyFromObject(dnsPolicy))),
+		Namespace:     dnsPolicy.Namespace,
+	}); err != nil {
+		return nil, err
+	}
+
+	return slice.MapErr(list.Items, func(obj v1alpha1.DNSHealthCheckProbe) (*v1alpha1.DNSHealthCheckProbe, error) {
+		return &obj, nil
+	})
+}
+
+func getProbesForEndpoint(endpoint *v1alpha1.Endpoint, probes []*v1alpha1.DNSHealthCheckProbe) []*v1alpha1.DNSHealthCheckProbe {
+	retProbes := []*v1alpha1.DNSHealthCheckProbe{}
+	for _, probe := range probes {
+		for _, target := range endpoint.Targets {
+			if strings.Contains(probe.Name, target) {
+				retProbes = append(retProbes, probe)
+			}
+		}
+	}
+	return retProbes
+}
```

Comment:
What endpoints do we current create probes for? Is it just the ultimate targets (the gateway address) or all endpoints?

Reply:
it's for anything that resolves outside of the DNS Record (i.e. CNAMEs to other zones, or A records)

Reply:
though the more general answer is: yes, the gateway addresses

Reply:
We create probes for all endpoints, i see: 
```
kubectl get DNSHealthCheckProbe -A
NAMESPACE                NAME                                                  HEALTHY   LAST CHECKED
multi-cluster-gateways   172.31.200.0-prod-web-api                             false     48s
multi-cluster-gateways   172.31.201.0-prod-web-api                             false     15s
multi-cluster-gateways   172.31.202.0-prod-web-api                             false     28s
multi-cluster-gateways   24osuu.lb-4ej5le.myapp.mn.hcpapps.net-prod-web-api    false     25s
multi-cluster-gateways   2c71gf.lb-4ej5le.myapp.mn.hcpapps.net-prod-web-api    false     45s
multi-cluster-gateways   default.lb-4ej5le.myapp.mn.hcpapps.net-prod-web-api   false     38s
multi-cluster-gateways   lb-4ej5le.myapp.mn.hcpapps.net-prod-web-api           false     35s
multi-cluster-gateways   lrnse3.lb-4ej5le.myapp.mn.hcpapps.net-prod-web-api    false     18s
```

So this logic we are adding is going to check the health of the CNAME records e.g. lb-4ej5le.myapp.mn.hcpapps.net-prod-web-api, if that is "unhealthy", remove that and also anything below it.

What scenario are we trying to cover by checking those explicitly? Is a health probe per gateway address (The IP or hostname being returned in the Gateway) not enough? I'm not saying it's wrong, I'm just trying to see what we are gaining by doing it this way.

A simpler solution that i was imagining was that we would use the HealthProbe information during the creation of the `dns.MultiClusterGatewayTarget`, that is passed into `setEndpoints`, rather than building all the endpoints and then removing them again. Essentially remove "unhealthy" cluster/gateway/addresses before it even gets to `setEndpoints`. That way health stuff doesn't need to worry about the record structure and CNAMES  and all the rest, just that the service is accessible on the expected clusters via the expected host.

Reply:
Currently main does create too many probes, but if you run on this branch you should only see probes for A records - or CNAMEs that point outside the zone. For example, I currently have a dns record like this:
```
endpoints:
  - dnsName: 24osuu.lb-4ej5le.myapp.pb.hcpapps.net
    recordTTL: 60
    recordType: A
    targets:
    - 172.31.202.0
  - dnsName: 2c71gf.lb-4ej5le.myapp.pb.hcpapps.net
    recordTTL: 60
    recordType: A
    targets:
    - 172.31.200.0
  - dnsName: default.lb-4ej5le.myapp.pb.hcpapps.net
    providerSpecific:
    - name: weight
      value: "120"
    recordTTL: 60
    recordType: CNAME
    setIdentifier: 24osuu.lb-4ej5le.myapp.pb.hcpapps.net
    targets:
    - 24osuu.lb-4ej5le.myapp.pb.hcpapps.net
  - dnsName: default.lb-4ej5le.myapp.pb.hcpapps.net
    providerSpecific:
    - name: weight
      value: "120"
    recordTTL: 60
    recordType: CNAME
    setIdentifier: 2c71gf.lb-4ej5le.myapp.pb.hcpapps.net
    targets:
    - 2c71gf.lb-4ej5le.myapp.pb.hcpapps.net
  - dnsName: default.lb-4ej5le.myapp.pb.hcpapps.net
    providerSpecific:
    - name: weight
      value: "120"
    recordTTL: 60
    recordType: CNAME
    setIdentifier: lrnse3.lb-4ej5le.myapp.pb.hcpapps.net
    targets:
    - lrnse3.lb-4ej5le.myapp.pb.hcpapps.net
  - dnsName: lb-4ej5le.myapp.pb.hcpapps.net
    providerSpecific:
    - name: geo-code
      value: '*'
    recordTTL: 300
    recordType: CNAME
    setIdentifier: default
    targets:
    - default.lb-4ej5le.myapp.pb.hcpapps.net
  - dnsName: lrnse3.lb-4ej5le.myapp.pb.hcpapps.net
    recordTTL: 60
    recordType: A
    targets:
    - 172.31.201.0
  - dnsName: myapp.pb.hcpapps.net
    recordTTL: 300
    recordType: CNAME
    targets:
    - lb-4ej5le.myapp.pb.hcpapps.net
```
And the following health probes:
```
NAMESPACE                NAME                        HEALTHY   LAST CHECKED
multi-cluster-gateways   172.31.200.0-prod-web-api   true      31s
multi-cluster-gateways   172.31.201.0-prod-web-api   true      26s
multi-cluster-gateways   172.31.202.0-prod-web-api   true      21s
```

Reply:
Yeah OK, i see. So we are no longer creating the ones for all the CNAME endpoints which is good as i couldn't see why we needed to.

Since that is the case, you could probably have done this more simply as described here https://github.com/Kuadrant/multicluster-gateway-controller/pull/412#discussion_r1315571223. If you think of it as removing unhealthy gateway addresses rather than endpoints in the DNSRecord. From what i can gather from the changes in   `pkg/controllers/dnspolicy/dnspolicy_healthchecks.go` this would fit in better anyway (process the gateway addresses not the DNSRecord). You wouldn't need to figure out what CNAMES you need to remove since it would just never have added it in the first place if the addresses are all empty for a cluster/gateway.

Up to you though, just a suggestion.

Reply:
Resolving this here as it will potentially be addressed in a seperate pr.